File: coverage.py
#!/bin/env pypy

"""
Tools for copy number analysis and visualization.

Usage:
  coverage grid <genome_file> <window_size> [-s N]
  coverage cds <bam_file> <gtf_file>
  coverage telomere <bam_file>
  coverage downsample <wig_file> <fold>
  coverage sum <wig_files>...
  coverage unbias logratios <wig_file>
  coverage median filter <wig_file> <window_size>
  coverage format igv <wig_file>

Options:
  -h --help       Show this screen.
  -s --step=N     Step size for window placement [default: window size / 2].
"""

from __future__ import print_function
import sys, docopt, re, os, math, tempfile, shutil, subprocess
from pypette import zopen, shell, revcomplement, info, error, shell_stdout
from pypette import Object
from sam import read_sam, ref_sequence_sizes

class WigTrack: pass

def read_fixed_wig(wig_path):
	tracks = {}
	chr = ''
	for line in zopen(wig_path):
		if line.startswith('fixedStep'):
			if chr:
				track.values = values[0:N]	 # Remove preallocated space
			track = Object()
			values = np.zeros(1000000)
			m = re.search(r'chrom=(\w+)', line)
			chr = m.group(1)
			m = re.search(r'start=(\d+)', line)
			track.start = int(m.group(1))
			m = re.search(r'step=(\d+)', line)
			track.step = int(m.group(1))
			m = re.search(r'span=(\d+)', line)
			track.span = int(m.group(1)) if m else -1
			N = 0
			tracks[chr] = track
			continue
		
		if chr:
			values[N] = float(line)
			N += 1
	
	if chr: track.values = values[0:N]	 # Remove preallocated space
	return tracks
		



def parse_wig_header(line):
	m = re.search(r'chrom=(\w+)', line)
	chr = m.group(1)
	m = re.search(r'start=(\d+)', line)
	start = int(m.group(1))
	m = re.search(r'step=(\d+)', line)
	step = int(m.group(1))
	m = re.search(r'span=(\d+)', line)
	span = int(m.group(1)) if m else -1
	return (chr, start, step, span)





	


#################
# COVERAGE GRID #
#################

def coverage_grid(genome_path, winsize, step):
	for line in zopen(genome_path):
		if not line.strip(): continue
		c = line.rstrip('\n').split('\t')
		chr, chr_len = c[0], int(c[1])
		start = 1
		while start + winsize < chr_len:
			print('%s\t%d\t%d' % (chr, start - 1, start + winsize - 1))
			start += step









################
# COVERAGE CDS #
################

def coverage_cds(bam_path, gtf_path):
	
	chr_sizes = ref_sequence_sizes(bam_path)
	
	info('Constructing a map of coding regions...')
	coding = {}
	for chr, size in chr_sizes.iteritems():
		coding[chr] = [False] * size
	for line in zopen(gtf_path):
		if line.startswith('#'): continue
		cols = line.split('\t')
		if cols[2] != 'CDS': continue
		if len(cols[0]) > 5: continue   # Ignore chromosomes other than chrXX
		if not cols[0] in coding: continue
		coding[cols[0]][int(cols[3])-1:int(cols[4])] = True
		
	info('Calculating a coverage histogram...')
	coverage_hist = [0] * 200
	chr = ''
	pos = 0
	for line in shell_stdout('bedtools genomecov -d -split -ibam %s' % bam_path):
		cols = line.split('\t')
		if cols[0] != chr:
			chr = cols[0]
			cds = coding[chr]
			pos = int(cols[1])-2
			info('%s...' % chr)
		pos += 1
		if cds[pos]:
			coverage_hist[min(int(cols[2]), len(coverage_hist)-1)] += 1
			
	print('Coverage histogram:')
	print('===================')
	for cov in range(0, len(coverage_hist)):
		print('%d: %d' % (cov, coverage_hist[cov]))
	





#####################
# COVERAGE TELOMERE #
#####################

def coverage_telomere(bam_path):
	
	# Method is based on "Assessing telomeric DNA content in pediatric cancers
	# using whole-genome sequencing data" by Parker et al.
	telo_seq = 'TTAGGG' * 4
	rev_telo_seq = revcomplement(telo_seq)
	
	telo_count = 0
	for al in read_sam(bam_path, 'u'):
		if telo_seq in al[9] or rev_telo_seq in al[9]: telo_count += 1
	
	print('%s\t%d' % (bam_path, telo_count))

	
		
	
		
		
#######################
# COVERAGE DOWNSAMPLE #
#######################
		
def coverage_downsample(wig_path, fold):
	wig, step = read_fixed_wig(wig_path)
	mid = (fold-1)/2
	for chr, data in wig.iteritems():
		for k in xrange(data.size / fold):
			#data[k] = np.median(data[k*fold:k*fold+fold])    # FIXME
			data[k] = sorted(data[k*fold:k*fold+fold])[mid]
		data = data[0:data.size/fold]
		print('fixedStep chrom=%s start=1 step=%d span=%d' %
			(chr, step*fold, step*fold))
		for v in data: print('%.2f' % v)





################
# COVERAGE SUM #
################
		
def coverage_sum(wig_paths):
	wigs = [read_fixed_wig(p) for p in wig_paths]
	for chr in wigs[0]:
		total = wigs[0][chr].values.copy()
		for wig in wigs[1:]: total += wig[chr].values
		span = wigs[0][chr].span
		print('fixedStep chrom=%s start=%d step=%d%s' % (chr,
			wigs[0][chr].start, wigs[0][chr].step,
			(' span=%d' % span) if span > 0 else ''))
		for v in total: print(v)








#############################
# COVERAGE UNBIAS LOGRATIOS #
#############################
		
def coverage_unbias_logratios(wig_path):
	wig = read_fixed_wig(wig_path)

	# Find mode for each chromosome
	modes = {}
	for chr in wig:
		bins = np.arange(-4.975, 4.976, .05)
		hist = np.zeros(len(bins))
		for val in wig[chr].values:
			if not -5 <= val < 5: continue
			hist[int((val + 5) // 0.05)] += 1
		modes[chr] = bins[np.argmax(hist)]

	mode = sorted(modes.values())[len(modes)/2]

	for chr in wig:
		values = wig[chr].values
		values -= mode
		span = wig[chr].span
		print('fixedStep chrom=%s start=%d step=%d%s' % (chr,
			wig[chr].start, wig[chr].step,
			(' span=%d' % span) if span > 0 else ''))
		for v in values: print(v)








##########################
# COVERAGE MEDIAN FILTER #
##########################
		
def coverage_median_filter(wig_path, win_size):
	wig = read_fixed_wig(wig_path)

	for chr in wig:
		orig = wig[chr].values


		bins = np.arange(-4.975, 4.976, .05)
		hist = np.zeros(len(bins))
		for val in wig[chr].values:
			if not -5 <= val < 5: continue
			hist[int((val + 5) // 0.05)] += 1
		modes[chr] = bins[np.argmax(hist)]

	mode = sorted(modes.values())[len(modes)/2]

	for chr in wig:
		values = wig[chr].values
		values -= mode
		span = wig[chr].span
		print('fixedStep chrom=%s start=%d step=%d%s' % (chr,
			wig[chr].start, wig[chr].step,
			(' span=%d' % span) if span > 0 else ''))
		for v in values: print(v)




	

#######################
# COVERAGE FORMAT IGV #
#######################

def coverage_format_igv(wig_path):
	for line in zopen(wig_path):
		if line == 'nan\n': print('NaN')
		elif line == '-inf\n': print('-1000')
		elif line == 'inf\n': print('1000')
		else: sys.stdout.write(line)





#######################
# COMMAND LINE PARSER #
#######################

if __name__ == '__main__':
	args = docopt.docopt(__doc__)
	if args['grid']:
		wsize = int(args['<window_size>'])
		step = wsize / 2
		if args['--step'].isdigit(): step = int(args['--step'])
		coverage_grid(args['<genome_file>'], wsize, step=step)
	elif args['cds']:
		coverage_cds(args['<bam_file>'], args['<gtf_file>'])
	elif args['telomere']:
		coverage_telomere(args['<bam_file>'])
	elif args['downsample']:
		coverage_downsample(args['<wig_file>'], int(args['<fold>']))
	elif args['sum']:
		coverage_sum(args['<wig_files>'])
	elif args['unbias'] and args['logratios']:
		coverage_unbias_logratios(args['<wig_file>'])
	elif args['median'] and args['filter']:
		coverage_median_filter(args['<wig_file>'], args['<window_size>'])
	elif args['format'] and args['igv']:
		coverage_format_igv(args['<wig_file>'])

File: coverage.jl
#!/bin/env julia

help = """
Tools for copy number analysis and visualization.

Usage:
  coverage tiled <bed_file> <grid_file>
  coverage cds <bam_file> <gtf_file>
  coverage logratio <test_wig> <ref_wig> <min_ref>

Options:
  -h --help       Show this screen.
  -s --step=N     Step size for window placement [default: window size / 2].
"""

using DocOpt, GZip

function ref_sequence_sizes(sam_path)
	ref_sizes = Dict{ASCIIString, Int64}()
	for line in eachline(`samtools view -H $(sam_path)`)
		m = match(r"@SQ\tSN:(\w+)\tLN:(\d+)", line)
		if m == nothing; continue; end
		ref_sizes[m.captures[1]] = int(m.captures[2])
	end
	return ref_sizes
end

function zopen(path)
	if path == "-"; return STDIN; end
	if ismatch(r"\.gz$", path)
		return GZip.open(path)
	else
		return open(path)
	end
end

info(msg) = println(STDERR, msg)
int(s::ASCIIString) = parse(Int, s)
int(s::SubString{ASCIIString}) = parse(Int, s)

type Chromosome
	name::ASCIIString
	first::Int32
	last::Int32
end

function coverage_tiled(bed_path, grid_path)
	# Read the karyotype, window size and step size from the grid file
	grid_file = zopen(grid_path)
	c = split(readline(grid_file), '\t')
	chr = Chromosome(c[1], int(c[2]) + 1, 0)
	chromosomes = [chr]
	winsize = int(c[3]) - int(c[2]); prev_pos = int(c[2])
	c = split(readline(grid_file), '\t')
	step = int(c[2]) - prev_pos; prev_pos = int(c[2])
	while !eof(grid_file)
		c = split(readline(grid_file), '\t')
		window = [int(c[2]), int(c[3])]
		@assert(window[2] - window[1] == winsize)
		if c[1] != chr.name
			chr.last = prev_pos
			chr = Chromosome(c[1], window[1] + 1, 0)
			push!(chromosomes, chr)
		else
			@assert(window[1] - prev_pos == step)
		end
		prev_pos = window[1]
	end
	chr.last = prev_pos
	win_overlap = winsize - step

	chr_list = join([replace(c.name, "chr", "") for c in chromosomes], ", ")
	info("Window size: $winsize")
	info("Step size: $step")

	bed_file = zopen(bed_path)
	cols = split(readline(bed_file), '\t')
	chr = filter(c -> c.name == cols[1], chromosomes)[1]
	cov = zeros(Int32, cld(chr.last - chr.first + 1, step))
	info("Analyzing chromosome $(chr.name)")
	while !eof(bed_file)
		cols = split(readline(bed_file), '\t')
		if cols[1] != chr.name
			for c in cov; println("$c"); end
			chr = filter(c -> c.name == cols[1], chromosomes)[1]
			cov = zeros(Int32, cld(chr.last - chr.first + 1, step))
			info("Analyzing chromosome $(chr.name)")
		end

		# Convert BED coordinates to 1-based inclusive genomic coordinates
		start = int(cols[2]) + 1
		stop = int(cols[3])

		# Convert coordinates to window indexes
		start = fld(start - chr.first - win_overlap, step) + 1
		start = max(start, 1)
		stop = cld(stop - chr.first, step) + 1
		#stop = min(stop, length(cov))   # FIXME: Unnecessary?

		cov[start:stop] += 1
	end
end

function coverage_cds(sam_path, gtf_path)
	chr_sizes = ref_sequence_sizes(sam_path)

	info("Constructing a map of coding regions...")
	coding = [chr => falses(size) for (chr, size) in chr_sizes]
	for line in eachline(zopen(gtf_path))
		if startswith(line, '#'); continue; end
		cols = split(line, '\t')
		if cols[3] != "CDS"; continue; end
		if length(cols[1]) > 5; continue; end
		ccds = get(coding, cols[1], nothing)
		if ccds == nothing; continue; end
		ccds[int(cols[4]):int(cols[5])] = true
	end

	info("Calculating a coverage histogram...")
	coverage = 0:200
	coverage_hist = zeros(length(coverage))
	chr = ""; pos = 0
	for line in eachline(`bedtools genomecov -d -split -ibam $(sam_path)`)
		cols = split(line, '\t')
		if cols[1] != chr
			chr = cols[1]
			pos = int(cols[2])
			ccds = coding[chr]
		else
			pos += 1
		end
		if ccds[pos]
			count = min(int(cols[3]), 200) + 1
			coverage_hist[count] += 1
		end
	end

	println("Coverage histogram:")
	println("===================")
	for (cov, count) in zip(coverage, coverage_hist)
		println("$(cov): $(count)")
	end
end


function coverage_logratio()
	for line in open(args["logratio"]["file"])

	end
end


args = docopt(help)
if args["tiled"]
	coverage_tiled(args["<bed_file>"], args["<grid_file>"])
elseif args["cds"]
	coverage_cds(args["<bam_file>"], args["<gtf_file>"])
elseif args["logratio"]
	coverage_logratio()
end
